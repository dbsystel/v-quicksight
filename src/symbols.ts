export const EmbeddingContextInjectionKey = Symbol(
  "The key to be used with inject when accessing the created embeddingContext for usage with the amazon-quicksight-embedding-sdk",
)
