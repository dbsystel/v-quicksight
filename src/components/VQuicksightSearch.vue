<script setup lang="ts">
import type { VQuicksightSearchContentOptions, VQuicksightFrameOptions } from '../types'
import type {
  QSearchContentOptions,
  QSearchExperience,
  EmbeddingContext,
  ExperienceFrameMetadata,
  FrameOptions,
  EmbeddingEvents
} from 'amazon-quicksight-embedding-sdk'
import { nanoid } from 'nanoid'
import type { Ref } from 'vue'
import { computed, inject, ref, watch } from 'vue'
import { EmbeddingContextInjectionKey } from '../symbols'

const props = withDefaults(
  defineProps<
    VQuicksightFrameOptions &
      VQuicksightSearchContentOptions & {
        container?: string | HTMLElement // override from FrameOptions and make it optional
        id?: string
        question?: string
      }
  >(),
  {
    width: '100%',
    height: '100%',
    withIframePlaceholder: false,
    hideTopicName: false,
    allowTopicSelection: true
  }
)

const emit = defineEmits<{
  (e: 'change', data: { changeEvent: EmbeddingEvents; metadata?: ExperienceFrameMetadata }): void
  (
    e: 'message',
    data: { messageEvent: EmbeddingEvents; experienceMetadata?: ExperienceFrameMetadata }
  ): void
}>()

const embeddingContext = inject<Ref<EmbeddingContext>>(EmbeddingContextInjectionKey)

const searchFrame = ref<QSearchExperience>()

const containerId = computed(() => props.id || `v-quicksight-search-${nanoid(6)}`)
const frameOptions = computed<FrameOptions>(() => {
  return {
    url: props.url,
    container: props.container || `#${containerId.value}`,
    height: props.height,
    width: props.width,
    className: props.className,
    withIframePlaceholder: props.withIframePlaceholder,
    onChange: (changeEvent, metadata) => {
      emit('change', { changeEvent, metadata })
    }
  }
})
const contentOptions = computed<QSearchContentOptions>(() => {
  return {
    hideTopicName: props.hideTopicName,
    theme: props.theme,
    allowTopicSelection: props.allowTopicSelection,
    onMessage: async (messageEvent, experienceMetadata) => {
      emit('message', { messageEvent, experienceMetadata })
    }
  }
})

async function embedSearch(ctx: EmbeddingContext) {
  searchFrame.value = await ctx.embedQSearchBar(frameOptions.value, contentOptions.value)
}

async function setQuestion(frame: QSearchExperience, question?: string) {
  if (question) {
    return await frame.setQuestion(question)
  } else {
    return await frame.close()
  }
}

watch(
  () => embeddingContext?.value,
  (newValue, oldValue) => {
    if (newValue && !oldValue) {
      embedSearch(newValue)
    }
  },
  { immediate: true }
)

watch(
  () => props.question,
  (newValue) => {
    if (searchFrame.value) {
      setQuestion(searchFrame.value, newValue)
    }
  },
  { immediate: true }
)
</script>

<template>
  <div class="v-quicksight-search" :id="containerId"></div>
</template>
