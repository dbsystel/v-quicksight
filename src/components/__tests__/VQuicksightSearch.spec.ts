import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { nextTick, ref } from 'vue'

import { EmbeddingContextInjectionKey } from '../../symbols'
import VQuicksightSearch from '../VQuicksightSearch.vue'

const frameOptions = {
  container: '#custom-el',
  width: '300px',
  height: '700px',
  withIframePlaceholder: true,
  className: 'my-custom-class'
}
const contentOptions = {
  hideTopicName: true,
  theme: 'my-theme',
  allowTopicSelection: false
}

describe('VQuicksightSearch', () => {
  function setupComponent(props: Record<string, any> = {}) {
    const searchFrame = {
      close: vi.fn().mockResolvedValue({ success: true }),
      setQuestion: vi.fn().mockResolvedValue({ success: true })
    }
    const embeddingContext = {
      embedQSearchBar: vi.fn().mockResolvedValue(searchFrame)
    }

    const component = mount(VQuicksightSearch, {
      global: {
        provide: {
          [EmbeddingContextInjectionKey]: ref(embeddingContext)
        }
      },
      props: {
        ...props,
        url: 'https://my-embed-url.example.org',
        id: 'v-quicksight-search-1'
      }
    })
    return { component, searchFrame, embeddingContext }
  }

  it('renders as expected', () => {
    const { component } = setupComponent()
    expect(component.element).toMatchSnapshot()
  })

  it('should call embedQSearchBar with default parameters', async () => {
    const { embeddingContext } = setupComponent()
    await nextTick()
    expect(embeddingContext.embedQSearchBar).toHaveBeenCalledOnce()
    expect(embeddingContext.embedQSearchBar).toHaveBeenCalledWith(
      {
        className: undefined,
        container: '#v-quicksight-search-1',
        height: '100%',
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org',
        width: '100%',
        withIframePlaceholder: false
      },
      {
        hideTopicName: false,
        allowTopicSelection: true,
        onMessage: expect.anything()
      }
    )
  })

  it('should call embedQSearchBar with customized parameters', () => {
    const { embeddingContext } = setupComponent({ ...frameOptions, ...contentOptions })
    expect(embeddingContext.embedQSearchBar).toHaveBeenCalledOnce()
    expect(embeddingContext.embedQSearchBar).toHaveBeenCalledWith(
      {
        ...frameOptions,
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org'
      },
      {
        ...contentOptions,
        onMessage: expect.anything()
      }
    )
  })

  it('should call setQuestion when the passed question name changes', async () => {
    const { searchFrame, component } = setupComponent()
    await nextTick() // wait until first setup is finished
    await component.setProps({ question: 'How many apples were sold?' })
    expect(searchFrame.setQuestion).toHaveBeenCalledOnce()
    expect(searchFrame.setQuestion).toHaveBeenCalledWith('How many apples were sold?')
  })

  it('should call close when the question changes to a nullish value', async () => {
    const { searchFrame, component } = setupComponent({ question: 'How many apples were sold?' })
    await nextTick() // wait until first setup is finished
    await component.setProps({ question: '' })
    expect(searchFrame.close).toHaveBeenCalled()
  })
})
