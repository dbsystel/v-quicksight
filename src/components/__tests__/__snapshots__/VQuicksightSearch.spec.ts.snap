// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VQuicksightSearch > renders as expected 1`] = `
<div
  class="v-quicksight-search"
  id="v-quicksight-search-1"
/>
`;
